proposals/2019-02-12-design-for-a-python-toolchain.md

```diff
@@ -0,0 +1,246 @@
+
+The new toolchain-related definitions are implemented in Starlark. A byproduct of this is that the provider type for `py_runtime` is exposed to Starlark. We also add to `py_runtime` an attribute for declaring whether it represents a Python 2 or Python 3 runtime.
+
+## Motivation
+
+The goal is to make the native Python rules use the toolchain framework to resolve the Python runtime. Advantages include:
+
+* allowing each `py_binary` to use a runtime suitable for its target platform
+
+* allowing Python 2 and Python 3 targets to run in the same build without [hacks](https://github.com/bazelbuild/bazel/issues/4815#issuecomment-460777113)
+
+* making it easier to run Python-related builds under remote execution
+
+* adding support for autodetection of available system Python runtimes, without requiring ad hoc rule logic
+
+* removing `--python_top` and `--python_path`
+
+* bringing Python in line with other rule sets and Bazel's best practices
+
+**Non-goal:** This work does not allow individual `py_binary`s to directly name a Python runtime to use. Instead, this information should be worked into either the configuration or a future toolchain constraint system. See the FAQ, below.
+
+## Design
+
+### New definitions
+
+A new [toolchain type](https://docs.bazel.build/versions/master/toolchains.html#writing-rules-that-use-toolchains) is created at `@bazel_tools//tools/python:toolchain_type`. This is the type for toolchains that provide a way to run Python code.
+
+Toolchain rules of this type are expected to return a [`ToolchainInfo`](https://docs.bazel.build/versions/master/skylark/lib/ToolchainInfo.html) with two fields, `py2_runtime` and `py3_runtime`, each of type `PyRuntimeInfo`. They are used for `PY2` and `PY3` binaries respectively.
+
+```python
+def _some_python_toolchain_impl(ctx):
+    ...
+    return [platform_common.ToolchainInfo(
+        py2_runtime = PyRuntimeInfo(...),
+        py3_runtime = PyRuntimeInfo(...))]
+```
+
+If either Python 2 or Python 3 is not provided by the toolchain, the corresponding field may be set to `None`. This is strongly discouraged, as it will prevent any target relying on that toolchain from using that version of Python. Toolchains that do use `None` here should be registered with lower priority than other toolchains, so that they are chosen only as a fallback.
+
+`PyRuntimeInfo` is a newly-exposed provider returned by the native [`py_runtime`](https://docs.bazel.build/versions/master/be/python.html#py_runtime) rule. It can be loaded from `@bazel_tools//tools/python:toolchain.bzl`. It has the following fields, each of which corresponds to an attribute on `py_runtime`. (The last one, `python_version`, is newly added in this doc.)
+
+* `interpreter_path`: An absolute filesystem path to a Python interpreter (or a script that launches a Python interpreter, forwarding any command-line arguments) available on the target platform. Must be `None` if `interpreter` is non-`None`.
```

Comment:
For Python it's actually the target platform, because it's an interpreter executed at runtime as opposed to a compiler executed at build time. Bazel's "building" of a py_binary is really just laying out the runfiles and generating the stub script that'll be the entry point to its execution later.

(In the case where you're running a tool written in Python on the execution platform of another target, that second target's execution platform is the Python binary's "target" platform, so the description is still accurate.)